File: taurus-protect-sdk-typescript/src/services/action-service.ts
/**
 * Action service for Taurus-PROTECT SDK.
 *
 * Provides methods for retrieving automated action configurations.
 * Actions allow automated workflows to be triggered based on specific conditions
 * such as balance thresholds.
 */

import { NotFoundError, ValidationError } from '../errors';
import type { ActionsApi } from '../internal/openapi/apis/ActionsApi';
import { actionEnvelopeFromDto, actionEnvelopesFromDto } from '../mappers/action';
import type { ActionEnvelope, ListActionsOptions } from '../models/action';
import {
  buildOffsetPagination,
  offsetRequest,
  type PaginatedResult,
} from '../models/pagination';
import { BaseService } from './base';
import { offsetQuery } from './paging';

/**
 * Service for managing automated actions in the Taurus-PROTECT system.
 *
 * Actions allow automated workflows to be triggered based on specific conditions
 * such as balance thresholds. When conditions are met, tasks like transfers or
 * notifications can be executed automatically.
 *
 * @example
 * ```typescript
 * // First page of actions
 * const { items, pagination } = await actionService.list();
 * for (const action of items) {
 *   console.log(`${action.id}: ${action.label} (${action.status})`);
 * }
 *
 * // Get a specific action
 * const action = await actionService.get('action-123');
 * console.log(`Action: ${action.label}, Auto-approve: ${action.autoApprove}`);
 *
 * // Next page
 * if (pagination.hasMore) {
 *   await actionService.list({ offset: pagination.nextOffset });
 * }
 * ```
 */
export class ActionService extends BaseService {
  private readonly actionsApi: ActionsApi;

  /**
   * Creates a new ActionService instance.
   *
   * @param actionsApi - The ActionsApi instance from the OpenAPI client
   */
  constructor(actionsApi: ActionsApi) {
    super();
    this.actionsApi = actionsApi;
  }

  /**
   * Lists a page of actions.
   *
   * @param options - Filters, `limit` (1-100, default 20) and `offset`
   * @returns The page of actions and its pagination, including the server's total
   * @throws {@link ValidationError} If limit or offset are out of bounds
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const page = await actionService.list({ limit: 10 });
   * for (const action of page.items) {
   *   console.log(`Action ${action.id}: ${action.label} (${action.status})`);
   * }
   * if (page.pagination.hasMore) {
   *   await actionService.list({ limit: 10, offset: page.pagination.nextOffset });
   * }
   * ```
   */
  async list(options?: ListActionsOptions): Promise<PaginatedResult<ActionEnvelope>> {
    const page = offsetRequest(options);

    return this.execute(async () => {
      const response = await this.actionsApi.actionServiceGetActions({
        ...offsetQuery(page),
        ids: options?.ids,
      });

      const rows = response.result ?? [];
      return {
        items: actionEnvelopesFromDto(rows),
        pagination: buildOffsetPagination('plus_rows', page, response, rows.length),
      };
    });
  }

  /**
   * Retrieves a specific action by its ID.
   *
   * @param actionId - The ID of the action to retrieve
   * @returns The action envelope
   * @throws {@link ValidationError} If actionId is empty
   * @throws {@link NotFoundError} If action is not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const action = await actionService.get('action-123');
   * console.log(`Action: ${action.label}`);
   * console.log(`Status: ${action.status}`);
   * console.log(`Auto-approve: ${action.autoApprove}`);
   * ```
   */
  async get(actionId: string): Promise<ActionEnvelope> {
    if (!actionId || actionId.trim() === '') {
      throw new ValidationError('actionId is required');
    }

    return this.execute(async () => {
      const response = await this.actionsApi.actionServiceGetAction({
        id: actionId,
      });

      const action = actionEnvelopeFromDto(response.action);

      if (!action) {
        throw new NotFoundError(`Action with id '${actionId}' not found`);
      }

      return action;
    });
  }
}

File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/service/ActionService.java
package com.taurushq.sdk.protect.client.service;

import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import com.taurushq.sdk.protect.client.mapper.ActionMapper;
import com.taurushq.sdk.protect.client.mapper.ApiExceptionMapper;
import com.taurushq.sdk.protect.client.model.ActionEnvelope;
import com.taurushq.sdk.protect.client.model.ActionResult;
import com.taurushq.sdk.protect.client.model.ApiException;
import com.taurushq.sdk.protect.client.model.Pagination;
import com.taurushq.sdk.protect.openapi.ApiClient;
import com.taurushq.sdk.protect.openapi.api.ActionsApi;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordActionEnvelope;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetActionReply;
import com.taurushq.sdk.protect.openapi.model.TgvalidatordGetActionsReply;

import java.util.Collections;
import java.util.List;

/**
 * Service for managing automated actions in the Taurus Protect system.
 * <p>
 * Actions allow automated workflows to be triggered based on specific conditions
 * such as balance thresholds. When conditions are met, tasks like transfers or
 * notifications can be executed automatically.
 * <p>
 * Example usage:
 * <pre>{@code
 * // First page of actions (default page size)
 * ActionResult actions = client.getActionService().getActions();
 *
 * // Get a specific action
 * ActionEnvelope action = client.getActionService().getAction("action-123");
 * }</pre>
 *
 * @see ActionEnvelope
 */
public class ActionService {

    private final ActionsApi actionsApi;
    private final ApiExceptionMapper apiExceptionMapper;

    /**
     * Creates a new ActionService.
     *
     * @param apiClient          the API client for making requests
     * @param apiExceptionMapper the mapper for converting API exceptions
     * @throws NullPointerException if any parameter is null
     */
    public ActionService(final ApiClient apiClient, final ApiExceptionMapper apiExceptionMapper) {
        Preconditions.checkNotNull(apiClient, "apiClient must not be null");
        Preconditions.checkNotNull(apiExceptionMapper, "apiExceptionMapper must not be null");
        this.actionsApi = new ActionsApi(apiClient);
        this.apiExceptionMapper = apiExceptionMapper;
    }

    /**
     * Retrieves the first page of actions, with the default page size.
     *
     * @return the actions and their pagination
     * @throws ApiException if the API call fails
     */
    public ActionResult getActions() throws ApiException {
        return getActions(0, 0, null);
    }

    /**
     * Retrieves a page of actions with optional filters.
     *
     * @param limit  the page size, 0 for the default ({@link Pagination#DEFAULT_PAGE_SIZE})
     * @param offset the offset, 0 for the first page
     * @param ids    optional list of action IDs to filter by
     * @return the actions and their pagination
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if limit or offset is out of range
     */
    public ActionResult getActions(final int limit, final long offset,
                                   final List<String> ids) throws ApiException {
        final int size = PagedOperation.ACTIONS.resolveSize("limit", limit);
        final long from = Pagination.resolveOffset("offset", offset);
        try {
            TgvalidatordGetActionsReply reply = actionsApi.actionServiceGetActions(
                    String.valueOf(size), from == 0 ? null : String.valueOf(from), ids);
            List<TgvalidatordActionEnvelope> rows = reply.getResult() == null
                    ? Collections.emptyList() : reply.getResult();
            return new ActionResult(ActionMapper.INSTANCE.fromDTOList(rows),
                    PagedOperation.ACTIONS.offsetPage(size, from, rows.size(), 0,
                            reply.getTotalItems(), null));
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }

    /**
     * Retrieves a specific action by its ID.
     *
     * @param actionId the ID of the action to retrieve
     * @return the action envelope
     * @throws ApiException             if the API call fails
     * @throws IllegalArgumentException if actionId is null or empty
     */
    public ActionEnvelope getAction(final String actionId) throws ApiException {
        Preconditions.checkArgument(!Strings.isNullOrEmpty(actionId), "actionId must not be null or empty");
        try {
            TgvalidatordGetActionReply reply = actionsApi.actionServiceGetAction(actionId);
            return ActionMapper.INSTANCE.fromDTO(reply.getAction());
        } catch (com.taurushq.sdk.protect.openapi.ApiException e) {
            throw apiExceptionMapper.toApiException(e);
        }
    }
}
